In [1]:
# Add the Pandas dependecy
import pandas as pd
import os

In [2]:
# Files to load
school_data_to_load = os.path.join("Resources", "schools_complete.csv")
student_data_to_load = os.path.join("Resources", "students_complete.csv")

school_data_df = pd.read_csv(school_data_to_load)
student_data_df = pd.read_csv(student_data_to_load)

In [3]:
school_data_df.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [4]:
student_data_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [5]:
# Determine if there are any missing values in the school data.
school_data_df.count()

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

In [6]:
# Determine if there are any missing values in the student data.
student_data_df.count()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

There are no missing values in any of the columns because the output shows 39,170 rows for each column in student_data_df. By feault, "null" values are not counted, so you can often quickly identify which columns have missing data.

In [7]:
# Determine if there are any missing values in the school data.
school_data_df.isnull()

School ID  school_name   type   size  budget
0       False        False  False  False   False
1       False        False  False  False   False
2       False        False  False  False   False
3       False        False  False  False   False
4       False        False  False  False   False
5       False        False  False  False   False
6       False        False  False  False   False
7       False        False  False  False   False
8       False        False  False  False   False
9       False        False  False  False   False
10      False        False  False  False   False
11      False        False  False  False   False
12      False        False  False  False   False
13      False        False  False  False   False
14      False        False  False  False   False

We see that every row that is not empty is given the Boolean value "False", which tells us that there are no missing values. This is another method to determine empty rows.

In [8]:
student_data_df.isnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0           False         False   False  False        False          False   
1           False         False   False  False        False          False   
2           False         False   False  False        False          False   
3           False         False   False  False        False          False   
4           False         False   False  False        False          False   
...           ...           ...     ...    ...          ...            ...   
39165       False         False   False  False        False          False   
39166       False         False   False  False        False          False   
39167       False         False   False  False        False          False   
39168       False         False   False  False        False          False   
39169       False         False   False  False        False          False   

       math_score  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
39165       False  
39166       False  
39167       False  
39168       False  
39169       False  

[39170 rows x 7 columns]

In [9]:
# To get the total number of empty rows, or rows that are "True", we can use the Pandas ".sum()" method
# after the ".isnull()" method.
student_data_df.isnull().sum()

Student ID       0
student_name     0
gender           0
grade            0
school_name      0
reading_score    0
math_score       0
dtype: int64

In [10]:
# The ".notnull()" method returns the opposite boolean value of an ".isnull()" method
student_data_df.notnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0            True          True    True   True         True           True   
1            True          True    True   True         True           True   
2            True          True    True   True         True           True   
3            True          True    True   True         True           True   
4            True          True    True   True         True           True   
...           ...           ...     ...    ...          ...            ...   
39165        True          True    True   True         True           True   
39166        True          True    True   True         True           True   
39167        True          True    True   True         True           True   
39168        True          True    True   True         True           True   
39169        True          True    True   True         True           True   

       math_score  
0            True  
1            True  
2            True  
3            True  
4            True  
...           ...  
39165        True  
39166        True  
39167        True  
39168        True  
39169        True  

[39170 rows x 7 columns]

In [11]:
student_data_df.notnull().sum()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

In [12]:
# Determine the data types included in the school DataFrame
school_data_df.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

In [13]:
# Determine the data types included in the student DataFrame
student_data_df.dtypes

Student ID        int64
student_name     object
gender           object
grade            object
school_name      object
reading_score     int64
math_score        int64
dtype: object

See cleaning_student_names.ipynb to see the process used to clean student name data

In [14]:
# Add each prefix and suffix to remove to a list.
prefixes_suffixes = ["Dr. ", "Mr. ","Ms. ", "Mrs. ", "Miss ", " MD", " DDS", " DVM", " PhD"]

# Iterate through the words in the "prefixes_suffixes" list and replace them with an empty space, "".
for word in prefixes_suffixes:
    student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")

In [15]:
student_data_df

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
3               3    Richard Scott      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  
0                 66          79  
1                 94          61  
2                 90          60  
3                 67          58  
4                 97          84  
...              ...         ...  
39165             99          90  
39166             95          70  
39167             73          84  
39168             99          90  
39169             95          75  

[39170 rows x 7 columns]

Now that we have cleaned the dataframes, its best practice to create a new dataframe to do calculations. This way, the original data is not affected. We will merge the two data frames on the shared column "school_name". There may be cases in which you want to merge on a column that has similar information in two seperate data frame, however they're name differently. Its goot practice to rename the columns so that they match. This will prevent duplicate columns are merging issues.

In [16]:
# Combine the data into a single dataset.
school_data_complete_df = pd.merge(student_data_df, school_data_df, on=["school_name", "school_name"])
school_data_complete_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
3          58          0  District  2917  1910635  
4          84          0  District  2917  1910635

In [18]:
# Get the total number of students
student_count = school_data_complete_df["Student ID"].count()
student_count

39170

In [23]:
# Calculate the total number of schools in school_data_df
school_count = school_data_df["school_name"].count()
print("There are " + str(school_count) + " number of schools in school_data_df")

# Calculate the total number of schools in school_data_complete_df
school_count2 = school_data_complete_df["school_name"].unique()
print("There are " + str(len(school_count2)) + " number of schools in school_data_complete_df")

There are 15 number of schools in school_data_df
There are 15 number of schools in school_data_complete_df


In [24]:
# Calculate total budget
total_budget = school_data_df["budget"].sum()
print("The total budget in the district is $" + str(total_budget))

The total budget in the district is $24649428


In [25]:
# Find average reading score
average_reading_score = school_data_complete_df["reading_score"].mean()
print("The average reading score is " + str(average_reading_score))

The average reading score is 81.87784018381414


In [26]:
# Find the average math score
average_math_score = school_data_complete_df["math_score"].mean()
print("The average math score is " + str(average_math_score))

The average math score is 78.98537145774827


We determine a passing score was 70 for both the reading and math assessments. To determine the percentage of students that passes both, we will find the total count of students who recieved >= 70, and divide that by the total number of students.

In [31]:
# Calculate the number of students passed the math assessment
passing_math = school_data_complete_df[school_data_complete_df["math_score"] >= 70]
passing_math_count = passing_math["student_name"].count()
passing_math_percent = passing_math_count/student_count * 100
print("A total of %" + str(passing_math_percent) + " passed the math exam")

A total of %74.9808526933878 passed the math exam


In [37]:
# Calculate the number of students who passed the reading assessment
passing_reading = school_data_complete_df[school_data_complete_df["reading_score"] >= 70]
passing_reading_count = passing_reading["student_name"].count()
passing_reading_percent = passing_reading_count/student_count * 100
print("A total of %" + str(passing_reading_percent) + " passed the reading exam")

A total of %85.80546336482001 passed the reading exam


In [39]:
# Calculate the Overall Passing Percentage
passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]
passing_math_reading_count = passing_math_reading["student_name"].count()
passing_math_reading_percent = passing_math_reading_count/student_count * 100
print("A total of %" + str(passing_math_reading_percent) + " passed both exams")

A total of %65.17232575950983 passed both exams


In [42]:
# Adding a list of values with keys to create a new DataFrame.
district_summary_df = pd.DataFrame(
          [{"Total Schools": school_count,
          "Total Students": student_count,
          "Total Budget": total_budget,
          "Average Math Score": average_math_score,
          "Average Reading Score": average_reading_score,
          "% Passing Math": passing_math_percent,
         "% Passing Reading": passing_reading_percent,
        "% Overall Passing": passing_math_reading_percent}])
district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

The Pandas ".map()" function is used for substituting each value in a Series with another value. Where the new value is generated from a function, a dictionary, or a series. This will be useful when formatting district_summary_df.

series.map("current_value_1" : "new_value_1",  "current_value_2" : "new_value_2", etc)

The Python ".format()" function is used to format a value to a specific format, such as one or two decimal places, or adding a thousands seperator.

"{value:format specification}".format(value)

In [43]:
# Format "Total Budget" to have the comma for a thousands separator, a decimal separator, and a "$".

district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("${:,.2f}".format)

district_summary_df["Total Budget"]

0    $24,649,428.00
Name: Total Budget, dtype: object

In [45]:
# Format the columns.
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.1f}".format)

district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.1f}".format)

district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.0f}".format)

district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.0f}".format)

district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.0f}".format)

In [46]:
# Reorder the columns in the order you want them to appear.
new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

# Assign district summary df the new column order.
district_summary_df = district_summary_df[new_column_order]
district_summary_df

Total Schools  Total Students    Total Budget Average Math Score  \
0             15           39170  $24,649,428.00               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65

This next task requires you to get the following key metrics for each school and place them in a school summary DataFrame. Here are the key metrics we're working with: 

School name,
School type,
Total students,
Total school budget,
Per student budget,
Average math score,
Average reading score,
% passing math,
% passing reading,
% overall passing

In [47]:
# Determine the school type.
per_school_types = school_data_df.set_index(["school_name"])["type"]
per_school_types

school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

In [48]:
# Add the per_school_types into a DataFrame for testing.
df = pd.DataFrame(per_school_types)
df

type
school_name                    
Huang High School      District
Figueroa High School   District
Shelton High School     Charter
Hernandez High School  District
Griffin High School     Charter
Wilson High School      Charter
Cabrera High School     Charter
Bailey High School     District
Holden High School      Charter
Pena High School        Charter
Wright High School      Charter
Rodriguez High School  District
Johnson High School    District
Ford High School       District
Thomas High School      Charter

In [50]:
# Calculate the total student count.
per_school_counts = school_data_df.set_index(["school_name"])["size"]
per_school_counts

school_name
Huang High School        2917
Figueroa High School     2949
Shelton High School      1761
Hernandez High School    4635
Griffin High School      1468
Wilson High School       2283
Cabrera High School      1858
Bailey High School       4976
Holden High School        427
Pena High School          962
Wright High School       1800
Rodriguez High School    3999
Johnson High School      4761
Ford High School         2739
Thomas High School       1635
Name: size, dtype: int64

The ".value_counts()" method will return a series of data with the number od times each school_name appears in a row.

In [51]:
# Calculate the total student count.
per_school_counts = school_data_complete_df["school_name"].value_counts()
per_school_counts

Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: school_name, dtype: int64

In [52]:
# Calculate the total school budget.
per_school_budget = school_data_df.set_index(["school_name"])["budget"]
per_school_budget

school_name
Huang High School        1910635
Figueroa High School     1884411
Shelton High School      1056600
Hernandez High School    3022020
Griffin High School       917500
Wilson High School       1319574
Cabrera High School      1081356
Bailey High School       3124928
Holden High School        248087
Pena High School          585858
Wright High School       1049400
Rodriguez High School    2547363
Johnson High School      3094650
Ford High School         1763916
Thomas High School       1043130
Name: budget, dtype: int64

In [53]:
# Calculate the per capita spending.
per_school_capita = per_school_budget / per_school_counts
per_school_capita

Bailey High School       628.0
Cabrera High School      582.0
Figueroa High School     639.0
Ford High School         644.0
Griffin High School      625.0
Hernandez High School    652.0
Holden High School       581.0
Huang High School        655.0
Johnson High School      650.0
Pena High School         609.0
Rodriguez High School    637.0
Shelton High School      600.0
Thomas High School       638.0
Wilson High School       578.0
Wright High School       583.0
dtype: float64

In [55]:
# Calculate the math scores.
student_school_math = student_data_df.set_index(["school_name"])["math_score"]
student_school_math

school_name
Huang High School     79
Huang High School     61
Huang High School     60
Huang High School     58
Huang High School     84
                      ..
Thomas High School    90
Thomas High School    70
Thomas High School    84
Thomas High School    90
Thomas High School    75
Name: math_score, Length: 39170, dtype: int64

In [57]:
# Calculate the average test scores.
per_school_math = school_data_complete_df.groupby(["school_name"]).mean()["math_score"]
per_school_math

school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64

In [58]:
per_school_reading = school_data_complete_df.groupby(["school_name"]).mean()["reading_score"]
per_school_reading

school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.725724
Thomas High School       83.848930
Wilson High School       83.989488
Wright High School       83.955000
Name: reading_score, dtype: float64

 To get the passing percentages, we need to:
 1. Determine what is the passing grade.
 2. Get the number of students who passed math and reading.
 3. Get the students who passed math and passed reading
 
Using the groupby() function will split an object (like a DataFrame) and apply a mathematical operation.

In [59]:
# Calculate the passing scores by creating a filtered DataFrame.
per_school_passing_math = school_data_complete_df[(school_data_complete_df["math_score"] >= 70)]

per_school_passing_reading = school_data_complete_df[(school_data_complete_df["reading_score"] >= 70)]

In [60]:
# Calculate the number of students passing math and passing reading by school.
per_school_passing_math = per_school_passing_math.groupby(["school_name"]).count()["student_name"]

per_school_passing_reading = per_school_passing_reading.groupby(["school_name"]).count()["student_name"]

In [61]:
# Calculate the percentage of passing math and reading scores per school.
per_school_passing_math = per_school_passing_math / per_school_counts * 100

per_school_passing_reading = per_school_passing_reading / per_school_counts * 100

In [62]:
# Calculate the students who passed both math and reading.
per_passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]

# Calculate the number of students who passed both math and reading.
per_passing_math_reading = per_passing_math_reading.groupby(["school_name"]).count()["student_name"]

# Calculate the overall passing percentage.
per_overall_passing_percentage = per_passing_math_reading / per_school_counts * 100

per_overall_passing_percentage

Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
dtype: float64

In [63]:
# Adding a list of values with keys to create a new DataFrame.

per_school_summary_df = pd.DataFrame({
             "School Type": per_school_types,
             "Total Students": per_school_counts,
             "Total School Budget": per_school_budget,
             "Per Student Budget": per_school_capita,
             "Average Math Score": per_school_math,
           "Average Reading Score": per_school_reading,
           "% Passing Math": per_school_passing_math,
           "% Passing Reading": per_school_passing_reading,
           "% Overall Passing": per_overall_passing_percentage})
per_school_summary_df.head()

School Type  Total Students  Total School Budget  \
Bailey High School      District            4976              3124928   
Cabrera High School      Charter            1858              1081356   
Figueroa High School    District            2949              1884411   
Ford High School        District            2739              1763916   
Griffin High School      Charter            1468               917500   

                      Per Student Budget  Average Math Score  \
Bailey High School                 628.0           77.048432   
Cabrera High School                582.0           83.061895   
Figueroa High School               639.0           76.711767   
Ford High School                   644.0           77.102592   
Griffin High School                625.0           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

Now to format the data frame!

In [64]:
# Format the Total School Budget and the Per Student Budget columns.
per_school_summary_df["Total School Budget"] = per_school_summary_df["Total School Budget"].map("${:,.2f}".format)
per_school_summary_df["Per Student Budget"] = per_school_summary_df["Per Student Budget"].map("${:,.2f}".format)

per_school_summary_df.head()

School Type  Total Students Total School Budget  \
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

In [65]:
# Reorder the columns in the order you want them to appear.
new_column_order = ["School Type", "Total Students", "Total School Budget", "Per Student Budget", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

# Assign district summary df the new column order.
per_school_summary_df = per_school_summary_df[new_column_order]

per_school_summary_df.head()

School Type  Total Students Total School Budget  \
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

Using the Pandas sort_values() function, we can sort a DataFrame or Series for a given text, index, or column that is passed within the parentheses. In addition, we can add the parameter ascending=False to sort from highest to lowest, or ascending=True to sort from lowest to highest the value referenced in the sort_values() function. If we don't add the ascending parameter, the default is ascending=True.

In [ ]:
# Sort and show top five schools.
top_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=False)

top_schools.head()